File: packages/sdk-go/core/userop/validation_data.go
// Package userop provides validation data utilities for EIP-4337.
package userop

import (
	"math/big"

	"github.com/ethereum/go-ethereum/common"
)

// ============================================================================
// Constants
// ============================================================================

// SigValidationFailed indicates signature validation failed (authorizer = 1).
var SigValidationFailed = big.NewInt(1)

// ValidityBlockRangeFlag is the bit 47 flag for block number mode.
// When set on validUntil or validAfter, the value is a block number, not a timestamp.
var ValidityBlockRangeFlag = new(big.Int).SetUint64(0x800_000_000_000)

// ValidityBlockRangeMask extracts the actual value (lower 47 bits).
var ValidityBlockRangeMask = new(big.Int).SetUint64(0x7ff_fff_fff_fff)

// uint48Mask is the mask for 6-byte (48-bit) values.
var uint48Mask = new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 48), big.NewInt(1))

// ============================================================================
// Types
// ============================================================================

// ValidationData represents parsed EIP-4337 validationData (uint256).
//
// Layout: | authorizer (20 bytes) | validUntil (6 bytes) | validAfter (6 bytes) |
type ValidationData struct {
	// Authorizer: zero address = valid, 0x01 = SIG_VALIDATION_FAILED, other = aggregator
	Authorizer common.Address

	// ValidUntil is the expiry time/block (0 = infinite).
	ValidUntil *big.Int

	// ValidAfter is the start time/block.
	ValidAfter *big.Int

	// ValidUntilIsBlockNumber is true if validUntil uses block number mode.
	ValidUntilIsBlockNumber bool

	// ValidAfterIsBlockNumber is true if validAfter uses block number mode.
	ValidAfterIsBlockNumber bool

	// IsSignatureFailed is true if authorizer = 1 (SIG_VALIDATION_FAILED).
	IsSignatureFailed bool
}

// ============================================================================
// Parse
// ============================================================================

// ParseValidationData parses a uint256 validationData into its components.
func ParseValidationData(data *big.Int) *ValidationData {
	if data == nil {
		data = big.NewInt(0)
	}

	// authorizer: top 20 bytes (bits 96..255)
	authorizerRaw := new(big.Int).Rsh(data, 96)

	// validUntil: next 6 bytes (bits 48..95)
	validUntilRaw := new(big.Int).And(new(big.Int).Rsh(data, 48), uint48Mask)

	// validAfter: bottom 6 bytes (bits 0..47)
	validAfterRaw := new(big.Int).And(data, uint48Mask)

	// Block number mode detection (bit 47)
	validUntilIsBlockNumber := new(big.Int).And(validUntilRaw, ValidityBlockRangeFlag).Sign() > 0
	validAfterIsBlockNumber := new(big.Int).And(validAfterRaw, ValidityBlockRangeFlag).Sign() > 0

	// Extract actual values (mask out bit 47 if block mode)
	validUntil := new(big.Int).Set(validUntilRaw)
	if validUntilIsBlockNumber {
		validUntil.And(validUntil, ValidityBlockRangeMask)
	}

	validAfter := new(big.Int).Set(validAfterRaw)
	if validAfterIsBlockNumber {
		validAfter.And(validAfter, ValidityBlockRangeMask)
	}

	// Authorizer interpretation
	isSigFailed := authorizerRaw.Cmp(SigValidationFailed) == 0
	authorizer := common.BigToAddress(authorizerRaw)

	return &ValidationData{
		Authorizer:              authorizer,
		ValidUntil:              validUntil,
		ValidAfter:              validAfter,
		ValidUntilIsBlockNumber: validUntilIsBlockNumber,
		ValidAfterIsBlockNumber: validAfterIsBlockNumber,
		IsSignatureFailed:       isSigFailed,
	}
}

// ============================================================================
// Pack
// ============================================================================

// PackValidationData packs validation data components into a uint256.
//
// Parameters:
//   - authorizer: 0 (valid), 1 (sig failed), or aggregator address
//   - validUntil: expiry time/block (0 = infinite)
//   - validAfter: start time/block
//   - useBlockNumbers: if true, sets bit 47 flag on both validUntil and validAfter
func PackValidationData(authorizer common.Address, validUntil, validAfter *big.Int, useBlockNumbers bool) *big.Int {
	until := new(big.Int).Set(validUntil)
	after := new(big.Int).Set(validAfter)

	if useBlockNumbers {
		until.Or(until, ValidityBlockRangeFlag)
		after.Or(after, ValidityBlockRangeFlag)
	}

	// Pack: authorizer(20B) << 96 | validUntil(6B) << 48 | validAfter(6B)
	result := new(big.Int).SetBytes(authorizer.Bytes())
	result.Lsh(result, 96)
	result.Or(result, new(big.Int).Lsh(new(big.Int).And(until, uint48Mask), 48))
	result.Or(result, new(big.Int).And(after, uint48Mask))

	return result
}

// ============================================================================
// Helpers
// ============================================================================

// IsSignatureValidationFailed checks if validationData indicates signature failure.
func IsSignatureValidationFailed(data *big.Int) bool {
	authorizer := new(big.Int).Rsh(data, 96)
	return authorizer.Cmp(SigValidationFailed) == 0
}

// UsesBlockNumberMode checks if validationData uses block number mode for either field.
func UsesBlockNumberMode(data *big.Int) bool {
	validUntilRaw := new(big.Int).And(new(big.Int).Rsh(data, 48), uint48Mask)
	validAfterRaw := new(big.Int).And(data, uint48Mask)
	return new(big.Int).And(validUntilRaw, ValidityBlockRangeFlag).Sign() > 0 ||
		new(big.Int).And(validAfterRaw, ValidityBlockRangeFlag).Sign() > 0
}

File: packages/sdk-ts/core/src/utils/validationData.ts
/**
 * EIP-4337 validationData Utilities
 *
 * Parses and packs the `validationData` uint256 returned by
 * `validateUserOp` and `validatePaymasterUserOp`.
 *
 * Layout: | authorizer (20 bytes) | validUntil (6 bytes) | validAfter (6 bytes) |
 *
 * Block Number Mode (EIP-4337 v0.9):
 *   When bit 47 (0x800000000000) is set on validUntil or validAfter,
 *   the value represents a block number instead of a timestamp.
 */

import type { Address, Hex } from 'viem'

// ============================================================================
// Constants
// ============================================================================

/** Signature validation failed (authorizer = 1) */
export const SIG_VALIDATION_FAILED = 1n

/**
 * Block Number Mode flag (bit 47).
 * When set on validUntil or validAfter, the value is a block number, not a timestamp.
 */
export const VALIDITY_BLOCK_RANGE_FLAG = 0x800_000_000_000n

/** Mask to extract the actual value (lower 47 bits) */
export const VALIDITY_BLOCK_RANGE_MASK = 0x7ff_fff_fff_fffn

// ============================================================================
// Types
// ============================================================================

export interface ValidationData {
  /** 0 = valid, 1 = SIG_VALIDATION_FAILED, other = aggregator address */
  authorizer: Address | 0n | 1n
  /** Expiry time/block. 0 = infinite */
  validUntil: bigint
  /** Start time/block */
  validAfter: bigint
  /** true if validUntil uses block number mode */
  validUntilIsBlockNumber: boolean
  /** true if validAfter uses block number mode */
  validAfterIsBlockNumber: boolean
}

// ============================================================================
// Parse
// ============================================================================

/**
 * Parse a uint256 validationData into its components.
 *
 * @param validationData - The packed uint256 from validateUserOp/validatePaymasterUserOp
 */
export function parseValidationData(validationData: bigint): ValidationData {
  // authorizer: top 20 bytes (bits 96..255)
  const authorizerRaw = validationData >> 96n
  // validUntil: next 6 bytes (bits 48..95)
  const validUntilRaw = (validationData >> 48n) & 0xffff_ffff_ffffn
  // validAfter: bottom 6 bytes (bits 0..47)
  const validAfterRaw = validationData & 0xffff_ffff_ffffn

  // Block number mode detection (bit 47)
  const validUntilIsBlockNumber = (validUntilRaw & VALIDITY_BLOCK_RANGE_FLAG) !== 0n
  const validAfterIsBlockNumber = (validAfterRaw & VALIDITY_BLOCK_RANGE_FLAG) !== 0n

  // Extract actual values (mask out bit 47 if block mode)
  const validUntil = validUntilIsBlockNumber
    ? validUntilRaw & VALIDITY_BLOCK_RANGE_MASK
    : validUntilRaw
  const validAfter = validAfterIsBlockNumber
    ? validAfterRaw & VALIDITY_BLOCK_RANGE_MASK
    : validAfterRaw

  // Authorizer interpretation
  let authorizer: Address | 0n | 1n
  if (authorizerRaw === 0n) {
    authorizer = 0n
  } else if (authorizerRaw === 1n) {
    authorizer = 1n
  } else {
    authorizer = (`0x${authorizerRaw.toString(16).padStart(40, '0')}` as Address)
  }

  return {
    authorizer,
    validUntil,
    validAfter,
    validUntilIsBlockNumber,
    validAfterIsBlockNumber,
  }
}

// ============================================================================
// Pack
// ============================================================================

/**
 * Pack validation data components into a uint256.
 *
 * @param authorizer - 0 (valid), 1 (sig failed), or aggregator address
 * @param validUntil - Expiry time/block (0 = infinite)
 * @param validAfter - Start time/block
 * @param useBlockNumbers - If true, sets bit 47 flag on both validUntil and validAfter
 */
export function packValidationData(
  authorizer: bigint | Address,
  validUntil: bigint,
  validAfter: bigint,
  useBlockNumbers: boolean = false
): bigint {
  let authorizerValue: bigint
  if (typeof authorizer === 'string') {
    authorizerValue = BigInt(authorizer)
  } else {
    authorizerValue = authorizer
  }

  let until = validUntil
  let after = validAfter

  if (useBlockNumbers) {
    until = until | VALIDITY_BLOCK_RANGE_FLAG
    after = after | VALIDITY_BLOCK_RANGE_FLAG
  }

  return (
    (authorizerValue << 96n) |
    ((until & 0xffff_ffff_ffffn) << 48n) |
    (after & 0xffff_ffff_ffffn)
  )
}

// ============================================================================
// Helpers
// ============================================================================

/** Check if validationData indicates signature failure */
export function isSignatureValidationFailed(validationData: bigint): boolean {
  const authorizer = validationData >> 96n
  return authorizer === SIG_VALIDATION_FAILED
}

/** Check if validationData uses block number mode for either field */
export function usesBlockNumberMode(validationData: bigint): boolean {
  const validUntilRaw = (validationData >> 48n) & 0xffff_ffff_ffffn
  const validAfterRaw = validationData & 0xffff_ffff_ffffn
  return (
    (validUntilRaw & VALIDITY_BLOCK_RANGE_FLAG) !== 0n ||
    (validAfterRaw & VALIDITY_BLOCK_RANGE_FLAG) !== 0n
  )
}
